week9/answers.sql

```diff
@@ -1,3 +1,167 @@
--- Create your tables, views, functions and procedures here!
+DROP SCHEMA IF EXISTS social;
 CREATE SCHEMA social;
 USE social;
+
+CREATE TABLE users (
+	user_id INT UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+   first_name VARCHAR(255),
+   last_name VARCHAR(255),
+   email VARCHAR(255),
+   created_on DATETIME DEFAULT CURRENT_TIMESTAMP
+);
+
+CREATE TABLE sessions (
+	session_id INT UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+   user_id INT UNSIGNED NOT NULL,
+   created_on DATETIME DEFAULT CURRENT_TIMESTAMP,
+   updated_on DATETIME DEFAULT CURRENT_TIMESTAMP,
+   FOREIGN KEY (user_id) REFERENCES users(user_id)
+   	ON UPDATE CASCADE
+   	ON DELETE CASCADE
+);
+
+CREATE TABLE friends (
+	user_friend_id INT UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+	user_id INT UNSIGNED NOT NULL,
+	friend_id INT UNSIGNED NOT NULL,
+	FOREIGN KEY (user_id) REFERENCES users(user_id)
+		ON UPDATE CASCADE
+		ON DELETE CASCADE,
+	FOREIGN KEY (friend_id) REFERENCES users(user_id)
+		ON UPDATE CASCADE
+		ON DELETE CASCADE
+);
+
+CREATE TABLE posts (
+	post_id INT UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+	user_id INT UNSIGNED NOT NULL,
+	created_on DATETIME DEFAULT CURRENT_TIMESTAMP,
+	updated_on DATETIME DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+	content MEDIUMTEXT
+);
+
+CREATE TABLE notifications (
+	notification_id INT UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+	user_id INT UNSIGNED NOT NULL,
+	post_id INT UNSIGNED NOT NULL,
+	FOREIGN KEY (user_id) REFERENCES users(user_id)
+		ON UPDATE CASCADE
+		ON DELETE CASCADE,
+	FOREIGN KEY (post_id) REFERENCES posts(post_id)
+		ON UPDATE CASCADE
+		ON DELETE CASCADE
+);
+
+CREATE OR REPLACE VIEW notification_posts AS (
+	SELECT
+		notif.user_id,
+		users.first_name,
+		users.last_name,
+		posts.post_id,
+		posts.content
+	FROM
+		posts
+		INNER JOIN notifications AS notif
+			ON posts.post_id = notif.post_id
+		INNER JOIN users
+			ON posts.user_id = users.user_id
+);
+
+
+DELIMITER $$
+CREATE PROCEDURE new_user_notif(IN new_id INT UNSIGNED, IN new_fname VARCHAR(255), IN new_lname VARCHAR(255))
+BEGIN
+	
+	DECLARE id INT UNSIGNED;
+	
+	DECLARE end_of_cursor TINYINT DEFAULT FALSE;
+	DECLARE user_cursor CURSOR FOR
+		SELECT user_id FROM users WHERE user_id != new_id;
+	DECLARE CONTINUE HANDLER FOR NOT FOUND
+		SET end_of_cursor = TRUE;
+		
+	INSERT INTO posts
+		(user_id, content)
+	VALUES
+		(new_id, CONCAT(new_fname, ' ', new_lname, ' just joined!'));
+		
+	OPEN user_cursor;
+	user_loop : LOOP
+	
+		FETCH user_cursor INTO id;
+		IF end_of_cursor THEN
+			LEAVE user_loop;
+		END IF;
+		
+		INSERT INTO notifications
+			(user_id, post_id)
+		VALUES
+			(id, (SELECT MAX(post_id) FROM posts));
```

Comment:
You'll find that this doesn't scale with parallel operations (aka, you create a post on line 83, meanwhile another user somewhere in the world also creates a post before this line is executed... you will be providing notifications to other user's private posts due to a concurrency issue). Checkout https://wwcc.instructure.com/courses/2267209/assignments/27957588?module_item_id=68821305 for how to safely perform this operation.